File: python/libertyclient.py
# libertyclient - code for performing basic actions with the Repeated Measures liberty socket server
# e.g. set the sample rate, units, etc.

import constant
import time
import socketclient

# SETMODE - set the mode to binary or ASCII
#
# setMode(socket,mode)
# 0 = ASCII
# 1 = binary
def setmode(s, mode):
    tosend = "%c,%d" % (constant.LIBERTY_SetMode, mode)
    s.send(tosend.encode())

# samplerate can be 120Hz or 240Hz
def setsamplerate(s, samplerate):
    if samplerate == 120:
        code = 3
    elif samplerate == 240:
        code = 4
    else:
        print("Sample rate must be 120 Hz or 240 Hz - not updated")
        return

    tosend = "%c,%d" % (constant.LIBERTY_SetSampleRate, code)
    s.send(tosend.encode())

# SETHEMISPHERE - set the active hemisphere
# sethemisphere(socket, sensor, hemisphere[3])
# e.g. for sensor 0
# sethemisphere(s, 0, [0, -1, 0]); # =Y is the active hemipshere for sensor 0

def sethemisphere(s, sensor, hemisphere):
    tosend = "%c,%d,%d,%d,%d" % (constant.LIBERTY_SetHemisphere, sensor, hemisphere[0], hemisphere[1], hemisphere[2])
    s.send(tosend.encode())

# setunits - set the units to cm (1) or inches (0)
def setunits(s, units):
    tosend = "%c,%d" % (constant.LIBERTY_SetUnits, units)
    s.send(tosend.encode())

# RESETFRAMECOUNT - reset the frame count
def resetframecount(s):
    tosend = "%c" % constant.LIBERTY_ResetFrameCount
    s.send(tosend.encode())

# SETOUTPUTFORMAT - set the output format
# 0 = just position, 1 = also orientation
def setoutputformat(s, sensor, recordorientation):
    if recordorientation == 1:
        format = 2
    else:
        format = 1

    tosend = "%c,%d,%d" % (constant.LIBERTY_SetOutputFormat, sensor, format)
    s.send(tosend.encode())

# SETALIGNMENTFRAME - set the alignment frame (which way is x,y,z)
#
# setalignmentframe(s,sensor,coords)
#
# coords should be 9 integers, with the position of the origin (3
# numbers), the direction of the x axis (3 numbers) and the direction of
# the y axis (3 numbers).
#
# e.g. setalignmentframe(s,0,[0,0,0,0,1,0,0,0,-1])
def setalignmentframe(s, sensor, alignmentframe):
    tosend = "%c,%d,%d,%d,%d,%d,%d,%d,%d,%d,%d" % (constant.LIBERTY_AlignmentReferenceFrame, sensor,
                                                   alignmentframe[0], alignmentframe[1], alignmentframe[2],
                                                   alignmentframe[3], alignmentframe[4], alignmentframe[5],
                                                   alignmentframe[6], alignmentframe[7], alignmentframe[8])

    s.send(tosend.encode())

# Get a single sample - data size will be numsensors*6+1
def getsinglesample(s, numsensors):
    tosend = "%c,%d" % (constant.LIBERTY_GetSingleSample, numsensors)
    s.send(tosend.encode())

    data = s.recv(1000)
    decoded = data.decode("utf-8")
    result = list(map(float,decoded.split(",")))
    return result

def getupdaterate(s):
    tosend = "%c" % constant.LIBERTY_GetUpdateRate
    s.send(tosend.encode())

    data = s.recv(10)
    rawrate = int(data)

    if rawrate == 3:
        return 120
    elif rawrate == 4:
        return 240
    else:
        return -1

# Perform standard setup for the device
# setupdevice(s, nummarkers, samplerate, hemisphere, alignmentframe)
# hemisphere should have 3 elements, e.g. [0 1 0] for the +y hemisphere
# alignment frame should have 9 numbers: [origin(3) x direction (3) y direction (3)]
def setupdevice(s, nummarkers, samplerate, hemisphere, alignmentframe):

    setmode(s, 1)  # Set to binary
    time.sleep(0.5)
    setsamplerate(s, samplerate)  # sample rate to 240 Hz
    time.sleep(0.5)
    # Get update rate to make sure liberty is still answering
    if getupdaterate(s) != samplerate:
        print("Problem after setsamplerate\n")
        return -1

    setunits(s, 1)  # 1 = cm
    time.sleep(0.5)
    # Get update rate to make sure liberty is still answering
    if getupdaterate(s) != samplerate:
        print("Problem after setunits\n")
        return -1

    # set the active hemisphere
    for k in range(nummarkers):
        sethemisphere(s, k, hemisphere)
        time.sleep(0.1)
        if getupdaterate(s) != samplerate:
            print("Problem after sethemisphere\n")
            return -1

    # Reset the frame count
    resetframecount(s)
    time.sleep(0.3)
    if getupdaterate(s) != samplerate:
        print("Problem after resetframecount\n")
        return -1

    # set the output format
    for k in range(nummarkers):
        setoutputformat(s, k, 1)
        time.sleep(0.1)
        if getupdaterate(s) != samplerate:
            print("Problem after setoutputformat\n")
            return -1

    # set the alignment reference frame (which way are origin (3 numbers),
    # x axis(3 numbers), y axis(3 numbers))
    for k in range(nummarkers):
        setalignmentframe(s, k, alignmentframe)
        time.sleep(0.1)
        if getupdaterate(s) != samplerate:
            print("Problem after setalignmentframe\n")
            return -1

    # get a sample to check everything is working
    result = getsinglesample(s, nummarkers)
    if result[0] == 0:
        print("Did not receive a result from the server - check everything is on and connected")
        return -1

    print("Received sample from liberty: timestamp " + str(result[0]) + "\n")
    for k in range(nummarkers):
        todisplay = "Sensor %d: %.3f, %.3f, %.3f, %.3f, %.3f, %.3f" % (k, result[k * 6 + 1], result[k * 6 + 2],
                                                                         result[k * 6 + 3], result[k * 6 + 4],
                                                                         result[k * 6 + 5], result[k * 6 + 6])
        print(todisplay)

    return 0

File: c++/LibertyClient/LibertyClient.cpp
#include "LibertyClient.h"
#include "messagecodes.h"

int setupdevice(SOCKET ConnectSocket, int numSensors) {

	int samplerate = 240;

	// Set the mode to binary
	printf("Setting mode to binary\n");
	setMode(ConnectSocket,1);
	Sleep(300);

	// Set the sample rate to 240 Hz
	printf("Setting sample rate to 240 Hz\n");
	setsamplerate(ConnectSocket,samplerate);
	Sleep(300);

	//We get the update rate as a way of checking that the liberty is still
	// answering us (and not stuck running some command or error)

	if(getupdaterate(ConnectSocket) != samplerate) {
		printf("Problem after setsamplerate\n");
		return -1;
	}

	// set the units to cm
	printf("Setting units to cm\n");
	setunits(ConnectSocket,1);
	Sleep(300);
	if(getupdaterate(ConnectSocket) != samplerate) {
		printf("Problem after setunits\n");
		return -1;
	}

	// set the active hemisphere
	for (int k=0;k<numSensors;k++) {
		printf("Setting hemisphere for %d to 0,1,0\n",k);
		sethemisphere(ConnectSocket,k,0,1,0);
		Sleep(100);
		if(getupdaterate(ConnectSocket) != samplerate) {
		printf("Problem after sethemisphere\n");
		return -1;
		}
	}

	// Reset the frame count
	printf("Resetting the frame count\n");
	resetframecount(ConnectSocket);
	Sleep(300);

	if(getupdaterate(ConnectSocket) != samplerate) {
		printf("Problem after resetframecount\n");
	}

	// set the output format
	for (int k=0;k<numSensors;k++) {
		setoutputformat(ConnectSocket,k,1);
		Sleep(100);
		if(getupdaterate(ConnectSocket) != samplerate) {
			printf("Problem after setoutputformat\n");
			return -1;
		}	
	}


	// set the alignment reference frame (which way are origin (3 numbers), x axis (3 numbers), y axis(3 numbers))
	for (int k=0;k<numSensors;k++) {
		setalignmentframe(ConnectSocket,k,
			0,0,0,1,0,0,0,1,0);
		Sleep(100);
		if(getupdaterate(ConnectSocket) != samplerate) {
			printf("Problem after setalignmentframe\n");
			return -1;
		}
	}

	// get a sample to check everything is working
	double result[500];
	getsinglesample(ConnectSocket,numSensors,result);
	if (result[0]==0) {
		printf("Did not receive a result from the server - check everything is on and connected");
		return -1;
	}

	printf("Received sample from liberty: timestamp: %f\n",result[0]);
	for (int k=0;k<numSensors;k++) {
		printf("Sensor %d: %.3f, %.3f, %.3f, %.3f, %.3f, %.3f\n",k,result[k*6+1],result[k*6+2],result[k*6+3],result[k*6+4],result[k*6+5],result[k*6+6]);
	}

	return 0;

}

// SETMODE - set the mode to binary or ASCII
//
// setmode(socket,mode)
// 0 = ASCII
// 1 = binary
void setMode(SOCKET ConnectSocket,int mode) {
	char message[50];
	sprintf(message,"%c,%d",LIBERTY_SetMode,mode);
	printf("Sending message: %s\n",message);
	sendMessage(ConnectSocket,message);
}

// SETHEMISPHERE - set the active hemisphere
// 
// sethemisphere(socket,hemisphere_x,hemisphere_y,hemisphere_z)
//
// e.g. sethemisphere(socket,0,0,0,1) % +Z is active hemisphere for sensor 0
// e.g. sethemisphere(socket,0,0,-1,0); % =Y is the active hemipshere for sensor 0
void sethemisphere(SOCKET ConnectSocket,int sensor,int hemisphere_x,int hemisphere_y,int hemisphere_z) {
	char message[50];
	sprintf(message,"%c,%d,%d,%d,%d",LIBERTY_SetHemisphere,sensor,hemisphere_x,hemisphere_y,hemisphere_z);
	printf("Sending message: %s\n",message);
	sendMessage(ConnectSocket,message);
}

// setsamplerate(socket, samplerate)
//
// samplerate can be 120Hz or 240Hz
void setsamplerate(SOCKET ConnectSocket,int samplerate) {
	int code = -1;
	if (samplerate==120)
		code = 3;
	else if (samplerate==240)
		code = 4;
	else {
		printf("Sample rate must be 120 Hz or 240 Hz - not updated");
		return;
	}

	char message[50];
	sprintf(message,"%c,%d",LIBERTY_SetSampleRate,code);
	printf("Sending message: %s\n",message);
	sendMessage(ConnectSocket,message);
}

// SETUNITS - set the units to cm or inches
//
// setunits(l,units)
// 0 = inches
// 1 = cm
void setunits(SOCKET ConnectSocket,int units) {
	char message[50];
	sprintf(message,"%c,%d",LIBERTY_SetUnits,units);
	printf("Sending message: %s\n",message);
	sendMessage(ConnectSocket,message);
}

// RESETFRAMECOUNT - reset the frame count
void resetframecount(SOCKET ConnectSocket) {
	char message[50];
	sprintf(message,"%c",LIBERTY_ResetFrameCount);
	printf("Sending message: %s\n",message);
	sendMessage(ConnectSocket,message);
}

// SETOUTPUTFORMAT - set the output format
//
// setoutputformat(socket,sensor,recordOrientation)
// recordOrientation should be 0 or 1

void setoutputformat(SOCKET ConnectSocket,int sensor,int recordOrientation) {
	int format;
	if (recordOrientation==1)
		format = 2;
	else
		format = 1;

	char message[50];
	sprintf(message,"%c,%d,%d",LIBERTY_SetOutputFormat,sensor,format);
	printf("Sending message: %s\n",message);
	sendMessage(ConnectSocket,message);
}

// SETALIGNMENTFRAME - set the alignment frame (which way is x,y,z)
//
// setalignmentframe(socket,sensor,coords)
//
// coords should be 9 integers, with the position of the origin (3
// numbers), the direction of the x axis (3 numbers) and the direction of
// the y axis (3 numbers). 
//
// e.g. setalignmentframe(socket,0,0,0,0,0,1,0,0,0,-1)
void setalignmentframe(SOCKET ConnectSocket,int sensor,
	int alignmentframe_0,int alignmentframe_1,int alignmentframe_2,
	int alignmentframe_3,int alignmentframe_4,int alignmentframe_5,
	int alignmentframe_6,int alignmentframe_7,int alignmentframe_8) {

		char message[100];
		sprintf(message,"%c,%d,%d,%d,%d,%d,%d,%d,%d,%d,%d",LIBERTY_AlignmentReferenceFrame,sensor,
			alignmentframe_0,alignmentframe_1,alignmentframe_2,
			alignmentframe_3,alignmentframe_4,alignmentframe_5,
			alignmentframe_6,alignmentframe_7,alignmentframe_8);

		printf("Sending message: %s\n",message);
		sendMessage(ConnectSocket,message);
}

// Get a single sample - data size will be numsensors*6+1
void getsinglesample(SOCKET ConnectSocket,int numsensors, double *result) {

	char message[50];
	sprintf(message,"%c,%d",LIBERTY_GetSingleSample,numsensors);
	printf("Sending message: %s\n",message);
	sendMessage(ConnectSocket,message);

	char buffer[1000];
	receiveMessage(ConnectSocket, buffer, 1000);

	// Parse the result
	int i=-1;
	char* remaining = strtok(buffer, ",");
	while (remaining != NULL) {
		i++;
		result[i] = atof(remaining);		
		remaining = strtok(NULL, ",");
	}
}

int getupdaterate(SOCKET ConnectSocket) {

	char message[50];
	sprintf(message,"%c",LIBERTY_GetUpdateRate);
	sendMessage(ConnectSocket,message);

	char buffer[10];
	receiveMessage(ConnectSocket,buffer,10);
	int rawrate = atoi(buffer);

	if (rawrate == 3)
		return 120;
	else if (rawrate == 4)
		return 240;
	else
		return -1;

}